File: MATLAB/Evaluation.m
%% ////////////////////////////////////////////////////////////////////////
%
% Title       : evaluation.m
%
% Description : evaluation of simulation
%
% Version     : 1.0
%
% Date        : 
%
% Input-Data  :  all structures
%
% Output-Data :  specific plots
%
%%//////////////////////////////////////////////////////////////////////////

function [] = evaluation(t,...
                         Aerodynamic,...
                         Airframe,...
                         Autopilot,...
                         AircraftData,...
                         Guidance)

%//////////////////////////////////////////////////////////////////////////
% trajcetory and velocity
%//////////////////////////////////////////////////////////////////////////
h1 = figure('Name','Trajcetory');
subplot(311)
plot3(Airframe.posNED(1,1:t),Airframe.posNED(2,1:t),-Airframe.posNED(3,1:t),'r',...
       Guidance.s_x(1,1:t),Guidance.s_y(1,1:t),-Guidance.s_z(1,1:t),'k--');
grid on
xlabel('X_{NED}')
ylabel('Y_{NED}')
zlabel('Z_{NED}')
legend('Trajectory','commanded Trajectory')
title('3D Trajectory')

subplot(312)
plot(1:t,Airframe.posNED(3,1:t),'r',...
      1:t,Guidance.s_z(1,1:t),'k--')
grid on
xlabel('Time [s]')
ylabel('Z_{NED}')
axis auto
legend('Trajectory','commanded Trajectory','Location','eastoutside')
title('Trajectory x-z-plane')

subplot(313)
plot(1:t,Airframe.posNED(2,1:t),'r',...
      1:t,Guidance.s_y(1,1:t),'k--')
legend('Trajectory','commanded Trajectory','Location','eastoutside')
grid on
xlabel('X_{NED}')
ylabel('Y_{NED}')
axis auto
title('Trajectory x-y-plane')

%//////////////////////////////////////////////////////////////////////////
% pitch axis parameter over time
%//////////////////////////////////////////////////////////////////////////
h2 = figure('Name','Pitch Axis');
subplot(511)
plot(1:t,Aerodynamic.Alpha(1,1:t),'r')
xlabel('Time [s]')
ylabel('\alpha [°]')
grid on
title('angle of attack')

subplot(512)
plot(1:t,Airframe.EulerAngles(2,1:t)*180/pi,'r')
xlabel('Time [s]')
ylabel('\theta [°]')
grid on
title('pitch angle')

subplot(513)
plot(1:t,Airframe.Gamma(1,1:t)*180/pi,'r')
xlabel('Time [s]')
ylabel('\gamma [°]')
grid on
title('flight path angle')

subplot(514)
plot(1:t,Airframe.rotRateBody(2,1:t),'r')
grid on
xlabel('Time [s]')
ylabel('q [rad/s]')
grid on
title('pitch rate')

subplot(515)
plot(1:t,Airframe.Eta(1,1:t)*180/pi,'r')
xlabel('Time [s]')
ylabel('\eta [°]')
grid on
title('elevator angle')

%//////////////////////////////////////////////////////////////////////////
% roll axis parameter over time
%//////////////////////////////////////////////////////////////////////////
 h3 = figure('Name','Roll Axis');

subplot(311)
plot(1:t,Airframe.EulerAngles(1,1:t)*180/pi,'r')
xlabel('Time [s]')
ylabel('\phi [°]')
grid on
title('roll angle')

subplot(312)
plot(1:t,Airframe.Xi(1,1:t)*180/pi,'r')
grid on
xlabel('Time [s]')
ylabel('\xi [°]')
title('aileron angle')

subplot(313)
plot(1:t,Airframe.rotRateBody(1,1:t),'r')
xlabel('Time [s]')
ylabel('p [rad/s]')
grid on
title('roll rate')

%//////////////////////////////////////////////////////////////////////////
% yaw axis parameter over time
%//////////////////////////////////////////////////////////////////////////
h4 = figure('Name','Yaw axis');
subplot(311)
plot(1:t,Aerodynamic.Beta(1,1:t)*180/pi,'r',...
      1:t,Guidance.Beta_com(1,1:t)*180/pi,'k--')
xlabel('Time [s]')
ylabel('\beta [°]')
grid on
legend('\beta','\beta_{com}','location','best')
title('angle of sideslip')

subplot(312)
plot(1:t,Airframe.Zeta(1,1:t)*180/pi,'r')
grid on
xlabel('Time [s]')
ylabel('\zeta [°]')
title('rudder angle')

subplot(313)
plot(1:t,Airframe.rotRateBody(3,1:t),'r')
grid on
xlabel('Time [s]')
ylabel('r [rad/s]')
title('yaw rate ')

%//////////////////////////////////////////////////////////////////////////
% velocity and thrust stick position over time
%//////////////////////////////////////////////////////////////////////////

h5 = figure('Name' , 'Velocity');
subplot(211)
plot(1:t,Airframe.StickPos(1,1:t)*100,'r')
grid on
title('Thrust Stick Position \delta ')
xlabel('Time [s]')
ylabel('\delta [%]')
title('Thrust Stick Position')


subplot(212)
plot(1:t,Airframe.V_absolut(1,1:t),'r',...
     1:t,Guidance.Velocity_com(1,1:t),'k--')
xlabel('Time [s]')
ylabel('V [m/s]')
title('absolut Velocity')
grid on
title('Velocity')
legend('V','V_{com}','Location','best')

h6 = figure('Name','G-Force');
subplot(311)
plot(1:t,Aerodynamic.A(1,1:t)/(AircraftData.mass*9.80665),'r')
grid on
title('G-Force')
xlabel('Time [s]')
ylabel('N')

subplot(312)
plot(1:t,Airframe.EulerAngles(1,1:t)*180/pi,'r',...
    1:t,Guidance.Phi_com(1,1:t)*180/pi,'k--')
grid on
xlabel('Time [s]')
ylabel('\phi [°]')
title('Roll Angle')
legend('\phi','\phi_{com}','Location','best')


subplot(313)
plot(1:t,Airframe.EulerAngles(2,1:t)*180/pi,'r',...
    1:t,Guidance.Theta_com(1,1:t)*180/pi,'k--')
grid on
xlabel('Time [s]')
ylabel('\theta')
title('Pitch Angle')
legend('\theta','\theta_{com}','Location','best')


figure('Name','Body Accelerations [m/s^2]')
subplot(311)
plot(1:t-1,Airframe.accBody(1,1:t-1),'r',...
     1:t-1,Guidance.accBody(1,1:t-1),'k--');
grid on
xlabel('Time [s]')
ylabel('a_x')
title('Acceleration x_{body}')
legend('real','BodyCom')

subplot(312)
plot(1:t-1,Airframe.accBody(2,1:t-1),'r',...
     1:t-1,Guidance.accBody(2,1:t-1),'k--');
grid on
xlabel('Time [s]')
ylabel('a_y')
title('Acceleration y_{body}')
legend('real','BodyCom')

subplot(313)
plot(1:t-1,Airframe.accBody(3,1:t-1),'r',...
     1:t-1,Guidance.accBody(3,1:t-1),'k--');
grid on
xlabel('Time [s]')
ylabel('a_z')
title('Acceleration z_{body}')
legend('real','BodyCom')

h15 = figure(251);
plot(1:t-1,Airframe.accBody(2,1:t-1)/9.80665,'r',...
     1:t-1,Guidance.accBody(2,1:t-1)/9.80665,'k--');
grid on
xlabel('Time [s]')
ylabel('n_y')
legend('real','BodyCom')
set(gca,'Fontsize',18)


h16 = figure(251);
plot(1:t-1,Airframe.accBody(3,1:t-1)/9.80665,'r',...
     1:t-1,Guidance.accBody(3,1:t-1)/9.80665,'k--');
grid on
xlabel('Time [s]')
ylabel('n_z')
legend('real','BodyCom')
set(gca,'Fontsize',18)

h17 = figure(250);
plot(1:t,Aerodynamic.Alpha(1,1:t),'r',...
     1:t,Guidance.alpha_com(1,1:t)*180/pi,'k--')
xlabel('Time [s]')
ylabel('\alpha [°]')
grid on
axis([0 t 3 7])
set(gca,'Fontsize',18)

print(h17,'-dpdf','-r1200','AlphaCOm.pdf'); 
print(h15,'-dpdf','-r1200','Ny.pdf'); 
print(h16,'-dpdf','-r1200','Nz.pdf'); 

end

File: Python/Evaluation.py
from matplotlib import pyplot as plt
import math
from matplotlib import rc
def readData(Filename):
    try:
        with open(Filename, "r") as myFile:
            # read in all lines ignoring Header
            File = myFile.readlines()[2::]

            rows = len(File)

            Data = [[] for i in range(len(File[0].split()))]

            # split data to columns
            for i in range(0, rows):
                temp = File[i].split()
                for ii in range(0, len(File[0].split())):
                    Data[ii].append(float(temp[ii]))
        return Data
    except:
        print("Kann "+Filename+" nicht öffnen")
        return None



if __name__ == "__main__":

    #read in tables
    Airframe = readData("../Output/AirframeData.txt")
    Guidance = readData("../Output/GuidanceData.txt")
    Thrust = readData("../Output/EngineData.txt")
    Aerodynamic = readData("../Output/AerodynamicData.txt")
    Navigation = readData("../Output/NavigationData.txt")
    Actuator = readData("../Output/ActuatorData.txt")
    IMU = readData("../Output/IMUData.txt")

    # show trajectory
    if Navigation != None:
        plt.figure("Trajectory")
        plt.subplot(211)
        plt.plot(Navigation[0], Navigation[4])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'Z$_{NED}$ [m]')
        plt.grid()
        plt.subplot(212)
        plt.plot(Navigation[2], Navigation[3])
        plt.xlabel(r'X$_{NED}$ [m]')
        plt.ylabel(r'Y$_{NED}$ [m]')
        plt.grid()
    else:
        print("Could not plot Trajectory")

    # plot longitudinal movement
    if Airframe and Aerodynamic and IMU and Navigation and Actuator != None:
        plt.figure("Longitudinal")
        plt.subplot(511)
        plt.plot(Airframe[0], Aerodynamic[1])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\alpha$ [rad]')
        plt.grid()
        plt.subplot(512)
        plt.plot(Airframe[0], Navigation[6])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\theta$ [rad]')
        plt.grid()
        plt.subplot(513)
        plt.plot(Airframe[0], IMU[1])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$q$ [rad/s]')
        plt.grid()
        plt.subplot(514)
        plt.plot(Airframe[0], Airframe[1])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\gamma$ [rad]')
        plt.grid()
        plt.subplot(515)
        plt.plot(Airframe[0], Actuator[0])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\eta$ [rad]')
        plt.grid()
    else:
        print("Could not plot longitudinal")

    #lateral movement
    if Airframe and Aerodynamic and IMU and Navigation and Actuator != None:
        plt.figure("lateral-roll")
        plt.subplot(311)
        plt.plot(Airframe[0], Navigation[5])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\phi$ [rad]')
        plt.grid()
        plt.subplot(312)
        plt.plot(Airframe[0], IMU[0])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$p$ [rad/s]')
        plt.grid()
        plt.subplot(313)
        plt.plot(Airframe[0], Actuator[1])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\xi$ [rad]')
        plt.grid()

        #lateral movement
        plt.figure("lateral-yaw")
        plt.subplot(511)
        plt.plot(Airframe[0], Aerodynamic[2])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\beta$ [rad]')
        plt.grid()
        plt.subplot(512)
        plt.plot(Airframe[0], Navigation[7])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\psi$ [rad]')
        plt.grid()
        plt.subplot(513)
        plt.plot(Airframe[0], IMU[2])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$r$ [rad]')
        plt.grid()
        plt.subplot(514)
        plt.plot(Airframe[0], Airframe[2])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\chi$ [rad]')
        plt.grid()
        plt.subplot(515)
        plt.plot(Airframe[0], Actuator[2])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\zeta$ [rad]')
        plt.grid()
    else:
        print("Could not plot lateral")

    if Navigation != None:
        plt.figure("Velocity")
        plt.subplot(211)
        plt.plot(Navigation[0], Navigation[1])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$V$ [m/s]')
        plt.axis([0,max(Navigation[0]),0,max(Navigation[1])+10])
        plt.grid()
        plt.subplot(212)
        plt.plot(Navigation[0], Actuator[3])
        plt.xlabel("Flighttime [s]")
        plt.ylabel(r'$\delta_{f}$ [-]')
        plt.grid()
    else:
        print("Could not plot Velocity")

    plt.show()